File: kmp/Pulsar/library/src/commonMain/kotlin/com/swmansion/pulsar/AdaptiveHaptics.kt
package com.swmansion.pulsar.kmp

sealed class AdaptivePresetConfig {
    class Function(val play: (PulsarPresets) -> Unit) : AdaptivePresetConfig()
    class Pattern(val pattern: PatternData) : AdaptivePresetConfig()
}

data class AdaptivePreset(
    val ios: AdaptivePresetConfig,
    val android: AdaptivePresetConfig,
)

class AdaptiveHaptics internal constructor(
    private val presets: PulsarPresets,
    private val composer: PatternComposer,
    private val config: AdaptivePresetConfig,
) {
    init {
        if (config is AdaptivePresetConfig.Pattern) {
            composer.parsePattern(config.pattern)
        }
    }

    fun play() {
        when (config) {
            is AdaptivePresetConfig.Function -> config.play(presets)
            is AdaptivePresetConfig.Pattern -> composer.play()
        }
    }

    fun stop() {
        if (config is AdaptivePresetConfig.Pattern) {
            composer.stop()
        }
    }
}

File: flutter/pulsar/lib/src/adaptive_haptics.dart
part of 'package:pulsar_haptics/pulsar.dart';

/// Plays platform-adaptive haptic feedback.
///
/// Created via [Pulsar.createAdaptiveHaptics]. Example:
/// ```dart
/// final pulsar = Pulsar();
/// final adaptive = await pulsar.createAdaptiveHaptics(AdaptivePreset(
///   ios: AdaptivePresetCallback(() => pulsar.presets.systemImpactMedium()),
///   android: AdaptivePresetPattern(myPattern),
/// ));
///
/// await adaptive.play();
/// // …
/// await adaptive.dispose();
/// ```
class AdaptiveHaptics {
  AdaptiveHaptics._({
    required PulsarPatternComposer composer,
    required AdaptivePresetConfig config,
  }) : _composer = composer,
       _config = config;

  final PulsarPatternComposer _composer;
  final AdaptivePresetConfig _config;

  /// Trigger the platform-appropriate haptic feedback.
  Future<void> play() async {
    switch (_config) {
      case AdaptivePresetCallback():
        await _config.play();
      case AdaptivePresetPattern():
        await _composer.play();
    }
  }

  /// Stop an in-progress pattern-based haptic. No-op for callback configs.
  Future<void> stop() async {
    if (_config is AdaptivePresetPattern) {
      await _composer.stop();
    }
  }

  /// Release the underlying native resources.
  Future<void> dispose() async => _composer.dispose();
}
